docs: onchain indices guide (#2298)
latticexyz/mud: docs/pages/guides/onchain-indices.mdx

pragma solidity >=0.8.24;

import { System } from "@latticexyz/world/src/System.sol";
import { Position } from "../codegen/index.sol";

contract MoveSystem is System {
  function move(int32 x, int32 y) public {
    address player = _msgSender();

    Position.set(player, x, y);
  }
}
```

To make the game more interesting, we would like to prevent players from moving to positions that are already occupied. However, it's not clear how to tell which players are at a given position onchain.

   <CollapseCode>

```solidity filename="MoveSystem.sol" copy showLineNumbers {11-14}
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.24;

import { System } from "@latticexyz/world/src/System.sol";
import { Position } from "../codegen/index.sol";

contract MoveSystem is System {
  function move(int32 x, int32 y) public {
    address player = _msgSender();

    bool positionIsEmpty = ?
    require(positionIsEmpty, "another player is at this position");

    Position.set(player, x, y);
  }
}
```

   </CollapseCode>

In other words, there is no reverse mapping - we can read the position of a given player, but not which players are at each position. It's time to add an index!

> There are two main ways to implement this - you can **create and manage your own indices**, or **use a module**.

Author: Interesting, I never thought about bold writing that way!

> ### Creating your own indices

Author: yesss I way prefer that. I was also thinking about something like "in-app indices"

> #### `PlayersAtPosition`

Reviewer: I think it would be best to remove this example. It's complicated and as you wrote in `PositionIsOccupied` example, needlessly so. My comments in this section are in case you decide to keep it.

Author: Agreed that having both an `address` and `bool` mapping is pointless!

One way to know whether a position is occupied is to have a list of which players are there, and check if that list is empty.

First, we declare an `PlayersAtPosition` table in our MUD config. The schema is essentially the reverse of `Position`; the key is an (x, y) coordinate and the value is an array of addresses currently at that position:

   <CollapseCode>
```ts filename="mud.config.ts" copy showLineNumbers {14-17}
import { mudConfig } from "@latticexyz/world/register";

export default mudConfig({
  tables: {
    Position: {
      keySchema: {
        player: "address",
      },
      valueSchema: {
        x: "int32",
        y: "int32",
      },
    },
    PlayersAtPosition: {
      keySchema: { x: "int32", y: "int32" },

>       valueSchema: { players: "address[]" },

Reviewer: ```suggestion
      valueSchema: { players: "address" },
```

We are changing the game so at most one player can be at a location. If we use `address` here, it makes the next code sample *a lot* simpler.

Author: Interesting point. I totally agree that the `filter` is needlessly confusing. 

The code will need to be modified a bit to check if this is zero but I will try it out!

Author: The main reason I implemented `address[]` was to compare the app-specific index with `keysWithValue` on more equal footing - the difference is gas would come from the hooks etc. But I guess that says something that we can optimize this index based on our app.

    },
  },
});
```

  </CollapseCode>

With our own indices, we need to be careful that they are maintained whenever the relevant tables are updated. Otherwise, they will not reflect current store data (for example, players in a `PlayersInPosition` list that don't actually have that `Position` value).

In this case, we need to add or remove addresses from `PlayersAtPosition` whenever a position is set. A good practice for this is to write a helper that atomically sets the table and index.

We can use a `setPosition` helper function to set both `Position` and the index.

> ```solidity filename="setPosition.sol" copy showLineNumbers

Reviewer: If you accept my suggestion above for using `player: address`, can you change this code to match? The `filter` function is non-trivial, IMAO.

// SPDX-License-Identifier: MIT
pragma solidity >=0.8.24;

import { Position, PositionData, PlayersAtPosition } from "../codegen/index.sol";

function filter(address[] memory arr, address element) pure returns (address[] memory) {
  address[] memory filtered = new address[](arr.length);
  uint256 filteredIndex;
  for (uint256 i; i < arr.length; i++) {
    if (arr[i] != element) {
      filtered[filteredIndex] = arr[i];
      filteredIndex++;
    }
  }

  // In-place update the length of the array
  // (Note: this does not update the free memory pointer)
  assembly {
    mstore(filtered, filteredIndex)

>   }

Reviewer 2: I'd add `/// @solidity memory-safe-assembly` to this assembly block, though it wouldn't matter much either way

```solidity filename="setPosition.sol" copy showLineNumbers
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.24;

import { Position, PositionData, PositionIsOccupied } from "../codegen/index.sol";

function setPosition(address player, int32 x, int32 y) {
  // Cache the player's previous position
  PositionData memory previousPosition = Position.get(player);

  Position.set(player, x, y);

  PositionIsOccupied.set(previousPosition.x, previousPosition.y, false);
  PositionIsOccupied.set(x, y, true);
}
```

Finally, to check whether the position is occupied, we simply read the index.

   <CollapseCode>

```solidity filename="MoveSystem.sol" copy showLineNumbers {5,11}
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.24;

import { System } from "@latticexyz/world/src/System.sol";
import { Position, PositionIsOccupied } from "../codegen/index.sol";

contract MoveSystem is System {
  function move(int32 x, int32 y) public {
    address player = _msgSender();

    bool positionIsEmpty = !PositionIsOccupied.get(x, y);
    require(positionIsEmpty, "another player is at this position");

    Position.set(player, x, y);
  }
}
```

   </CollapseCode>

> 

Reviewer: Add a note here that store hooks (linking to /store/store-hooks) can be used to make sure nobody sets a position without changing the index.

      root: true,
      args: [resolveTableId("Position")],
    },
  ],
}
```

  </CollapseCode>

Internally, this module uses [Store Hooks](https://mud.dev/store/store-hooks) to automatically update the index whenever a `Position` record is set.

After the module is installed, we can use a helper to get all addresses at a given position:

```solidity filename="MoveSystem.sol" copy showLineNumbers {4,6-7,13-18}
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.24;

import { PackedCounter } from "@latticexyz/store/src/PackedCounter.sol";
import { System } from "@latticexyz/world/src/System.sol";
import { getKeysWithValue } from "@latticexyz/world-modules/src/modules/keyswithvalue/getKeysWithValue.sol";
import { Position, PositionTableId } from "../codegen/index.sol";

contract MoveSystem is System {
  function move(int32 x, int32 y) public {
    address player = _msgSender();

    bool positionIsEmpty = getKeysWithValue(
      PositionTableId,
      Position.encodeStatic(x, y),
      PackedCounter.wrap(bytes32(0)),
      ""
    ).length == 0;
    require(positionIsEmpty, "another player is at this position");

    Position.set(player, x, y);
  }
}
```

> 

Reviewer: How about adding a 

<details>

<summary>Explanation</summary>

That explains the non-trivial code at 

```
    bool positionIsEmpty = getKeysWithValue(
      PositionTableId,
      Position.encodeStatic(x, y),
      PackedCounter.wrap(bytes32(0)),
      ""
    )
```
</details>

### Gas comparison

With MUD's [`GasReporter`](https://github.com/latticexyz/mud/tree/main/packages/gas-report), we can determine the gas use of `MoveSystem.move`:

```solidity filename="MoveSystem.t.sol" copy showLineNumbers
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.24;

import { MudTest } from "@latticexyz/world/test/MudTest.t.sol";
import { GasReporter } from "@latticexyz/gas-report/src/GasReporter.sol";

import { IWorld } from "../src/codegen/world/IWorld.sol";

contract MoveSystemTest is MudTest, GasReporter {
  function testMove() public {
    startGasReport("moving one player");
    IWorld(worldAddress).move(1, 1);
    endGasReport();
  }
}
```

> Comparing the three approaches, we find that managing our own indices is generally more gas-efficient, with the "position is occupied" index being the cheapest by far.

Reviewer: ```suggestion
Comparing the approaches, we find that managing our own index is generally more gas-efficient, whereas using a generic module to create the index is more expensive.
```

> #### `PlayersAtPosition`

Reviewer: If you decide to delete `PlayersAtPosition`, delete this one too. Otherwise, change as requested in the tables below.

```
┌───────────────────────┬──────────┬───────────────────┬──────────┐
│ File                  │ Test     │ Name              │ Gas used │
├───────────────────────┼──────────┼───────────────────┼──────────┤
│ test/MoveSystem.t.sol │ testMove │ moving one player │ 135922   │
└───────────────────────┴──────────┴───────────────────┴──────────┘
```

> 

Reviewer: ```suggestion
Note that the gas figures in this table are accurate as of writing. Future MUD versions are likely to be slightly different.
```

> #### `PositionIsOccupied`
> 
> ```
> ┌───────────────────────┬──────────┬───────────────────┬──────────┐
> │ File                  │ Test     │ Name              │ Gas used │
> ├───────────────────────┼──────────┼───────────────────┼──────────┤
> │ test/MoveSystem.t.sol │ testMove │ moving one player │ 64637    │
> └───────────────────────┴──────────┴───────────────────┴──────────┘
> ```
> 
> #### `KeysWithValue`
> 
> ```
> ┌───────────────────────┬──────────┬───────────────────┬──────────┐
> │ File                  │ Test     │ Name              │ Gas used │
> ├───────────────────────┼──────────┼───────────────────┼──────────┤
> │ test/MoveSystem.t.sol │ testMove │ moving one player │ 165000   │
> └───────────────────────┴──────────┴───────────────────┴──────────┘
> ```

Reviewer: ```suggestion
#### `PositionIsOccupied`

| Mechanism | Gas Used |
| ---------- | ---: |
| `PositionIsOccupied` | 64637 |
| `KeysWithValue` | 165000 |
```
